### PR TITLE
[Fix][meta_tool] meta tool build fix

### LINKED ISSUE #30781: [Bug] meta_tool build failure
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

master

### What's Wrong?

FAILED: src/tools/CMakeFiles/meta_tool.dir/meta_tool.cpp.o 
/home/rohit/opensource/ldb-toolchain/bin/clang++ -DBOOST_DATE_TIME_NO_LIB -DBOOST_DATE_TIME_POSIX_TIME_STD_CONFIG -DBOOST_STACKTRACE_USE_BACKTRACE -DBOOST_SYSTEM_NO_DEPRECATED -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX=1 -DBRPC_ENABLE_CPU_PROFILER -DENABLE_STACKTRACE -DHAVE_INTTYPES_H -DHAVE_NETINET_IN_H -DS2_USE_GFLAGS -DS2_USE_GLOG -DUSE_HADOOP_HDFS -DUSE_JEMALLOC -DUSE_MEM_TRACKER -DUSE_UNWIND -D__STDC_FORMAT_MACROS -Dmeta_tool_EXPORTS -I/opensource/doris/be/src/apache-orc/c++/include -I/opensource/doris/be/build_Debug/src/apache-orc/c++/include -I/opensource/doris/be/build_Debug/src/clucene/src/shared -I/opensource/doris/be/src/clucene/src/core -I/opensource/doris/be/src/clucene/src/shared -I/opensource/doris/be/src/clucene/src/contribs-lib -I/opensource/doris/be/src -I/opensource/doris/be/test -I/usr/lib/jvm/java-1.8.0-openjdk-amd64/include -I/usr/lib/jvm/java-1.8.0-openjdk-amd64/include/linux -isystem /opensource/doris/be/../gensrc/build -isystem /opensource/doris/thirdparty/installed/include -isystem /opensource/doris/thirdparty/installed/gperftools/include -isystem /opensource/doris/thirdparty/installed/include/dragonbox-1.1.3 -isystem /opensource/doris/thirdparty/installed/include/lz4 -isystem /opensource/doris/thirdparty/installed/include/zstd -O0  -g   -D OS_LINUX -g -Wall -Wextra -Werror -pthread -fstrict-aliasing -fno-omit-frame-pointer -Wnon-virtual-dtor -Wno-unused-parameter -Wno-sign-compare -fcolor-diagnostics -Wpedantic -Wshadow-field -Wunused -Wunused-command-line-argument -Wunused-exception-parameter -Wunused-volatile-lvalue -Wunused-template -Wunused-member-function -Wunused-macros -Wno-vla-extension -Wno-gnu-statement-expression -msse4.2 -mavx2 -gdwarf-aranges -std=gnu++20 -Winvalid-pch -Xclang -include-pch -Xclang /opensource/doris/be/build_Debug/CMakeFiles/pch.dir/cmake_pch.hxx.pch -Xclang -include -Xclang /opensource/doris/be/build_Debug/CMakeFiles/pch.dir/cmake_pch.hxx -MD -MT src/tools/CMakeFiles/meta_tool.dir/meta_tool.cpp.o -MF src/tools/CMakeFiles/meta_tool.dir/meta_tool.cpp.o.d -o src/tools/CMakeFiles/meta_tool.dir/meta_tool.cpp.o -c /opensource/doris/be/src/tools/meta_tool.cpp
/opensource/doris/be/src/tools/meta_tool.cpp:154:32: error: no matching constructor for initialization of 'DataDir'
            new (std::nothrow) DataDir(path.path, path.capacity_bytes, path.storage_medium));
                               ^       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opensource/doris/be/src/olap/data_dir.h:52:5: note: candidate constructor not viable: no known conversion from 'std::string' (aka 'basic_string<char>') to 'StorageEngine &' for 1st argument
    DataDir(StorageEngine& engine, const std::string& path, int64_t capacity_bytes = -1,
    ^
/opensource/doris/be/src/olap/data_dir.h:50:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 3 were provided
class DataDir {
      ^
1 error generated.

### What You Expected?

1. meta_tool build success
2. meta_tool run success

### How to Reproduce?

BUILD_TYPE=Debug bash build.sh --meta-tool --index-tool --be

### Anything Else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by rohitrs1983**
## Proposed changes

Issue Number: close #30781 

meta_tool build failure fix
[output.txt](https://github.com/apache/doris/files/14152207/output.txt)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/tools/meta_tool.cpp [hunk 1]
```diff
@@ -150,8 +152,19 @@ Status init_data_dir(const std::string& dir, std::unique_ptr<DataDir>* ret) {
...
-    std::unique_ptr<DataDir> p(
-            new (std::nothrow) DataDir(path.path, path.capacity_bytes, path.storage_medium));
+    doris::config::tablet_map_shard_size = 256;
+    doris::config::txn_map_shard_size = 1024;
+    doris::config::txn_shard_size = 1024;
+    doris::config::partition_disk_index_lru_size = 10000;
+    doris::config::delete_bitmap_agg_cache_capacity = 104857600;
+
+    doris::ExecEnv::GetInstance()->set_cache_manager(doris::CacheManager::create_global_instance());
+    doris::ExecEnv::GetInstance()->set_dummy_lru_cache(std::make_shared<doris::DummyLRUCache>());
```
Comment (platoneko) [inline be/src/tools/meta_tool.cpp:162]: Is it necessary to set cache for meta_tool?

File: be/src/tools/meta_tool.cpp [hunk 2]
```diff
@@ -150,8 +152,19 @@ Status init_data_dir(const std::string& dir, std::unique_ptr<DataDir>* ret) {
...
+
+    std::unique_ptr<doris::StorageEngine> _engine =
+            std::make_unique<doris::StorageEngine>(doris::EngineOptions {});
+    std::unique_ptr<DataDir> p(new (std::nothrow) DataDir(
```
Comment (platoneko) [inline be/src/tools/meta_tool.cpp:166]: The lifetime of the `StorageEngine` instance constructed in this way is only within this function body, but the `DataDir` instance still needs to be used after the function returns. Therefore, the `StorageEngine` instance encounters a "use after free" issue.

**Review by platoneko — CHANGES_REQUESTED** [review]
Please resolve the comment.

File: be/src/tools/meta_tool.cpp (hunk 1 above)
Comment (rohitrs1983) [reply be/src/tools/meta_tool.cpp]: fixed

File: be/src/tools/meta_tool.cpp (hunk 2 above)
Comment (rohitrs1983) [reply be/src/tools/meta_tool.cpp:168]: fixed

**Comment by Nitin-Kashyap** [comment]
run buildall

**Pull request closed without merging**